### PR TITLE
Add XML docs for exceptions that can be thrown by ModuleClient.SendEventAsync and DeviceClient.SendEventAsync

### LINKED ISSUE #2176: List of exceptions that can be thrown by ModuleClient.SendEventAsync is not complete
# Issue Transfer

This issue has been transferred from the Azure SDK for .NET repository, [#24191](https://github.com/Azure/azure-sdk-for-net/issues/24191).

## Details

The documentation mentions that the `SendEventAsync` method can throw an `OperationCanceledException`, and no other exception types are listed.
However, I have a situation where this method throws a `TimeoutException` as well:

image

It would be nice if the list of exceptions that can be thrown by this method is completed.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: dfaaeae3-f8de-02b1-7bc6-bf22027c0a99
* Version Independent ID: e48c3c7b-1944-6927-cefb-5e505dab8ce4
* Content: [ModuleClient.SendEventAsync Method (Microsoft.Azure.Devices.Client) - Azure for .NET Developers](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.devices.client.moduleclient.sendeventasync?view=azure-dotnet)
* Content Source: [xml/Microsoft.Azure.Devices.Client/ModuleClient.xml](https://github.com/Azure/azure-docs-sdk-dotnet/blob/master/xml/Microsoft.Azure.Devices.Client/ModuleClient.xml)
* Service: **iot-hub**
* GitHub Login: @rloutlaw
* Microsoft Alias: **routlaw**

### PR DESCRIPTION
**Pull request by Marusyk**
Added XML docs for exceptions that can be thrown by ModuleClient.SendEventAsync and DeviceClient.SendEventAsync
Please review.

Thank you in advance

## Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [x] This pull-request is submitted against the `master` branch.

## Reference/Link to the issue solved with this PR (if any)
Fixes #2176

### CODE REVIEW

**Comment by abhipsaMisra** [comment]
Thank you for submitting this PR. 
There are a couple of more exceptions that could be thrown from these APIs, would you mind adding the below test as well:

```csharp
        /// <exception cref="IotHubCommunicationException">Thrown if the client encounters a transient retryable exception. </exception>
        /// <exception cref="TimeoutException">Thrown if the service does not respond to the request within the timeout specified for the operation.
        /// The timeout values are largely transport protocol specific. Check the corresponding transport settings to see if they can be configured.
        /// The operation timeout for the client can be set using <see cref="OperationTimeoutInMilliseconds"/>.</exception>
        /// <exception cref="OperationCanceledException">Thrown if the service does not respond to the request before the expiration of the passed <see cref="CancellationToken"/>.
        /// If a cancellation token is not supplied to the operation call, a cancellation token with an expiration time of 4 minutes is used.
        /// </exception>
        /// <exception cref="SocketException">Thrown if a socket error occurs.</exception>
        /// <exception cref="WebSocketException">Thrown if an error occurs when performing an operation on a WebSocket connection.</exception>
        /// <exception cref="IOException">Thrown if an I/O error occurs.</exception>
        /// <exception cref="ClosedChannelException">Thrown if the MQTT transport layer closes unexpectedly.</exception>
        /// <exception cref="IotHubException">Thrown if an error occurs when communicating with IoT Hub service.
        /// If <see cref="IotHubException.IsTransient"/> is set to <c>true</c> then it is a transient exception.
        /// If <see cref="IotHubException.IsTransient"/> is set to <c>false</c> then it is a non-transient exception.</exception>
        /// <remarks>
        /// In case of a transient issue, retrying the operation should work. In case of a non-transient issue, inspect the error details and take steps accordingly.
        /// Please note that the above list is not exhaustive.
        /// </remarks>
```

**Comment by Marusyk** [comment]
@abhipsaMisra done
please review

File: iothub/device/src/DeviceClient.cs [hunk 1]
```diff
@@ -476,7 +476,21 @@ public void SetRetryPolicy(IRetryPolicy retryPolicy)
...
+        /// <exception cref="SocketException">Thrown if a socket error occurs.</exception>
+        /// <exception cref="WebSocketException">Thrown if an error occurs when performing an operation on a WebSocket connection.</exception>
+        /// <exception cref="IOException">Thrown if an I/O error occurs.</exception>
+        /// <exception cref="ClosedChannelException">Thrown if the MQTT transport layer closes unexpectedly.</exception>
+        /// <exception cref="IotHubException">Thrown if an error occurs when communicating with IoT Hub service.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>true</c> then it is a transient exception.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>false</c> then it is a non-transient exception.</exception>
+        /// <remarks>
+        /// In case of a transient issue, retrying the operation should work. In case of a non-transient issue, inspect the error details and take steps accordingly.
+        /// Please note that the above list is not exhaustive.
```
Comment (drwill-ms) [inline iothub/device/src/DeviceClient.cs:492]: I realize this text may have been supplied to you by Abhipsa, but I have a small issue with it.

It doesn't specify what list is not exhaustive, and when translated to the web documentation the list of exceptions are likely below the remarks section IIRC.

I suggest we reference the list directly.

```suggestion
        /// Please note that the list of exceptions is not exhaustive.
```

If you'd apply this change (or feel free to improve upon it) to this and all the other cases in this PR, that would be ideal.

**Review by drwill-ms — CHANGES_REQUESTED** [review]
One small wording change requested.

**Review by drwill-ms — APPROVED** [review]

**Comment by drwill-ms** [comment]
It looks like there is a build failure due to some of the types referenced in the doc comments not being found. You'll need to add some using statements up top to make sure the types are known.

**Comment by Marusyk** [comment]
I can`t see the build error because  - 401 - Uh-oh, you do not have access.

**Comment by drwill-ms** [comment]
> I can`t see the build error because - 401 - Uh-oh, you do not have access.

No problem. It is an internal pipeline. The error is a failed test due to service throttling. I'm rerunning the failed pipelines. I'm sure it'll go through and then I'll merge it.

It is looking good at this point. Thank you very much for your quick responses along the way.

**Pull request merged**